CHANGELOG.md

```diff
@@ -10,6 +10,7 @@
 - Fish is now more forgiving of missing or invalid $TERM values (#3850).
 - The `string` command now supports a `repeat` subcommand with the obvious behavior (#3864).
 - The `string match` command now supports a `--filter` flag to emit the entire string partially matched by a pattern (#3957).
+- The `string replace` command now supports a `--filter` flag to limit output to strings which underwent a replacement (#3348).
```

Comment:
I share your concern. On the other hand the two subcommands using this flag are very different in their respective behaviors. So while closely related the differing semantics for this flag is not unreasonable. If we deem the use of the same flag in both subcommands excessively confusing then the `string match --filter` case should be renamed. In the context of `string replace` the `--filter` flag has an obvious meaning whereas in the context of `string match` it does not.